### PR TITLE
Support `embed` package.

### PR DESCRIPTION
**Pull request by visualfc**
support go:embed
- gopherjs build/test/serve pkg
- gopherjs run/build simple.go

dep pkg [github.com/visualfc/goembed](https://github.com/visualfc/goembed)

Fixes https://github.com/gopherjs/gopherjs/issues/997

### CODE REVIEW

**Review by nevkontakte — CHANGES_REQUESTED** [review]
@visualfc thank you for this request, it is something I wished to do for a long time, although couldn't find time for.

I left a new minor comments, and I have a few more general questions and suggestions:

 - Have you attempted to measure how much more space the embedded data occupies in the compiled source compared to the original? Although this isn't something we have to do in this PR, investigating if we could compress the embedded data somehow would be a great improvement (from the issue #136 perspective especially).
 - To help me better understand how this PR works, could you post the generated `js_embed.go` source for some package? `embed/internal/embedtest` would be a good test subject.
 - How does gopherjs handle `//go:embed` instructions in files that do not import the `embed` package? Does it ignore them or error out? Is it the same as the official compiler?
 - A good first step to test this PR would be removing `embed/internal/embedtest` package from `.std_test_pkg_exclusions` and making sure the CI still passes.

**Comment by nevkontakte** [comment]
Also cc @flimzy who at some point planned to work on this feature, in case he has additional insights.

File: build/build.go [hunk 1]
```diff
@@ -548,12 +558,28 @@ func (s *Session) BuildFiles(filenames []string, pkgObj string, cwd string) erro
 		return fmt.Errorf("named files must all be in one directory; have: %v", strings.Join(dirList, ", "))
 	}
 
+	root := dirList[0]
+	ctx := build.Default
+	ctx.UseAllFiles = true
+	ctx.ReadDir = func(dir string) (infos []fs.FileInfo, err error) {
+		for _, f := range filenames {
```
Comment (flimzy) [inline build/build.go:565]: nit: `infos` could be preallocated here, if we remove the named return values.

File: build/embed.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	r := goembed.NewResolve()
+	var buf bytes.Buffer
+	buf.WriteString(strings.Replace(embed_head, "$pkg", bp.Name, 1))
+	buf.WriteString("\nvar (\n")
+	for _, v := range ems {
+		v.Spec.Names[0].Name = "_"
+		fs, _ := r.Load(bp.Dir, v)
+		switch v.Kind {
+		case goembed.EmbedBytes:
+			buf.WriteString(fmt.Sprintf("\t%v = []byte(%v)\n", v.Name, buildIdent(fs[0].Name)))
```
Comment (flimzy) [inline build/embed.go:58]: Is there a reason not to use `fmt.Fprintf` here?

File: build/embed.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+//go:linkname __gopherjs_embed_buildFS__ embed.buildFS
+func __gopherjs_embed_buildFS__(list []struct {
+	name string
+	data string
+	hash [16]byte
+}) (f embed.FS)
+`
+
+// checkEmbed is check package embed file
+func checkEmbed(bp *PackageData, fset *token.FileSet, files []*ast.File, test bool, xtest bool) (*ast.File, bool) {
```
[the remark this answers is not in the document]
Comment (flimzy) [reply build/embed.go:34]: It looks like the bool return value isn't really needed here, as a nil check on the *ast.File return value would accomplish the same.  If we do opt to keep the bool, let's document its purpose in the GoDoc comment.

File: compiler/natives/src/embed/embed.go [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+// +build js
+
+package embed
+
+func buildFS(list []struct {
+	name string
+	data string
+	hash [16]byte
+}) (f FS) {
+	var files []file
```
Comment (flimzy) [inline compiler/natives/src/embed/embed.go:11]: nit: This can be preallocated.

**Review by flimzy — COMMENTED** [review]
I'm really happy to see this PR!  Thanks for doing it!

I haven't tried running any of it yet, but left a couple comments inline, mostly stylistic.

It looks like all the heavy lifting is done in `github.com/visualfc/goembed`.  I see it's being used by at least a couple other packages, but I do wonder if it would make sense to copy that logic into gopherjs directly.  I'm curious of @nevkontakte 's thoughts.

I may be wrong, but I don't think we have any precedent for core compiler logic in an external package that doesn't come from upstream.  But maybe that's not important anyway.

**Comment by visualfc** [comment]
this PR current embed generate code on gopherjs/build used by ast. 
not support embed/internal/embedtest.TestAliases.
maybe move to gopherjs/compiler for used by types.Type to support aliases.
```
// +build go 1.17
var (
	//go:embed "testdata/hello.txt"
	helloT []T
)

type T byte
```

**Comment by nevkontakte** [comment]
> I may be wrong, but I don't think we have any precedent for core compiler logic in an external package that doesn't come from upstream. But maybe that's not important anyway.

@flimzy this is not entirely unprecedented, `github.com/neelance/astrewrite` or `github.com/shurcooL/httpfs`, which are pretty important to the project. So for the purposes of this PR, I don't mind adding `github.com/visualfc/goembed` as a dependency, although I am equally open to moving the core logic into the gopherjs repository as an internal package.

> this PR current embed generate code on gopherjs/build used by ast. not support embed/internal/embedtest.TestAliases. maybe move to gopherjs/compiler for used by types.Type to support aliases. remove `github.com/visualfc/goembed`. like linkname parse `//go:embed` to direct translate js pkg `init` code.
> 
> embed go1.16 : use ast embed go1.17 : use types for support aliases
> 
> ```
> // +build go 1.17
> var (
> 	//go:embed "testdata/hello.txt"
> 	helloT []T
> )
> 
> type T byte
> ```

Considering we are only targeting compatibility with a single Go version (1.18.x at this time), I think we only need to worry about Go 1.17+ behavior. That said, I am open to accepting this PR with TestAliases skipped, and then addressing it in a separate PR. Which ever way you prefer :)

**Comment by visualfc** [comment]
@nevkontakte 

refactoring `embedFiles` code and support alias string or []byte, but valid check by next types.check ( compiler ) .
- embedFiles rewrite ast for go:embed vars
- embedFiles generate js_embed.go only `__gopherjs_embed_buildFS__`  and go:embed const data
```
package main

import (
	"embed"
	_ "unsafe"
)

//go:linkname __gopherjs_embed_buildFS__ embed.buildFS
func __gopherjs_embed_buildFS__(list []struct {
	name string
	data string
	hash [16]byte
}) (f embed.FS)

const (
	__gopherjs_embed_74657374646174612f64617461312e747874__ = "\x68\x65\x6c\x6c\x6f\x20\x64\x61\x74\x61\x31"
	__gopherjs_embed_74657374646174612f64617461322e747874__ = "\x68\x65\x6c\x6c\x6f\x20\x64\x61\x74\x61\x32"
	__gopherjs_embed_74657374646174612f7375622f64617461312e747874__ = "\x73\x75\x62\x20\x64\x61\x74\x61\x31"
	__gopherjs_embed_74657374646174612f7375622f64617461322e747874__ = "\x73\x75\x62\x20\x64\x61\x74\x61\x32"
)
```

File: build/build.go (hunk 1 above)
Comment (visualfc) [reply build/build.go]: fixed

File: compiler/natives/src/embed/embed.go (hunk 4 above)
Comment (visualfc) [reply compiler/natives/src/embed/embed.go]: fixed

File: build/build.go [hunk 5]
```diff
@@ -290,6 +291,9 @@ func parseAndAugment(xctx XContext, pkg *PackageData, isTest bool, fileSet *toke
 	if errList != nil {
 		return nil, nil, errList
 	}
+	if f, found := checkEmbed(pkg, fileSet, files, isTest, isXTest); found {
+		files = append(files, f)
+	}
```
[the remark this answers is not in the document]
Comment (visualfc) [reply build/build.go]: now check error

File: build/build.go [hunk 6]
```diff
@@ -414,11 +422,13 @@ func (p *PackageData) TestPackage() *PackageData {
...
-			Dir:        p.Dir,
-			GoFiles:    p.XTestGoFiles,
-			Imports:    p.XTestImports,
+			Name:                 p.Name + "_test",
+			ImportPath:           p.ImportPath + "_test",
+			Dir:                  p.Dir,
+			GoFiles:              p.XTestGoFiles,
+			Imports:              p.XTestImports,
+			XTestEmbedPatterns:   p.XTestEmbedPatterns,
+			XTestEmbedPatternPos: p.XTestEmbedPatternPos,
```
[the remark this answers is not in the document]
Comment (visualfc) [reply build/build.go]: fixed

**Comment by nevkontakte** [comment]
@visualfc thank you for making these fixes. This week has been kind of busy for me so far, but I'll try to take a detailed look at your PR and so some additional testing in the next couple of days. Sorry about the delay.

File: build/embed.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+//go:linkname __gopherjs_embed_buildFS__ embed.buildFS
+func __gopherjs_embed_buildFS__(list []struct {
+	name string
+	data string
+	hash [16]byte
+}) (f embed.FS)
+`
+
+// embedFiles generates an additional source file, which initializes all variables in the package with a go:embed directive.
+func embedFiles(bp *PackageData, fset *token.FileSet, files []*ast.File) (*ast.File, error) {
```
Comment (nevkontakte) [inline build/embed.go:34]: nit: For consistency with the rest of the codebase, let's rename `bp` to `pkg`.

File: build/embed.go [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+		return nil, nil
+	}
+
+	ems, err := goembed.CheckEmbed(bp.EmbedPatternPos, fset, files)
+	if err != nil {
+		return nil, err
+	}
+
+	r := goembed.NewResolve()
+	for _, v := range ems {
```
Comment (nevkontakte) [inline build/embed.go:45]: nit: For consistency, let's rename `v` to `em`.

File: build/embed.go [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+
+// embedFiles generates an additional source file, which initializes all variables in the package with a go:embed directive.
+func embedFiles(bp *PackageData, fset *token.FileSet, files []*ast.File) (*ast.File, error) {
+	if len(bp.EmbedPatternPos) == 0 {
+		return nil, nil
+	}
+
+	ems, err := goembed.CheckEmbed(bp.EmbedPatternPos, fset, files)
+	if err != nil {
+		return nil, err
```
Comment (nevkontakte) [inline build/embed.go:41]:
```suggestion
		return nil, fmt.Errorf("failed to associate embed patterns with the AST nodes in package %q: %w", pkg.ImportPath, err)
```

**Review by nevkontakte — CHANGES_REQUESTED** [review]

**Comment by nevkontakte** [comment]
@visualfc I left what I think is the final round of comments, all of which a minor style suggestions.

I also familiarized myself with the with the `github.com/visualfc/goembed` and I think I mostly understand what's going on :) 

One thing to note though. https://github.com/visualfc/goembed/blob/main/fsys/fsys.go seems to be derived from https://github.com/golang/go/blob/master/src/cmd/go/internal/fsys/fsys.go. Would you mind adding the appropriate attribution notice to it, similar to what you have in https://github.com/visualfc/goembed/blob/master/resolve/resolve.go?

File: compiler/natives/src/embed/internal/embedtest/embed_test.go [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+// +build js
+
+package embedtest
+
+import (
+	"testing"
+)
+
+func TestUninitialized(t *testing.T) {
+	t.Skip("reflect.DeepEqual error for empty slice and variadic")
```
Comment (nevkontakte) [inline compiler/natives/src/embed/internal/embedtest/embed_test.go:11]:
```suggestion
	t.Skip("reflect.DeepEqual error for empty slice and variadic: https://github.com/gopherjs/gopherjs/issues/1147")
```

Comment (nevkontakte) [reply compiler/natives/src/embed/internal/embedtest/embed_test.go:11]: Actually, I sent https://github.com/gopherjs/gopherjs/pull/1150 to address the issue with reflect.DeepEqual(). As soon at it is merged, you should be able to re-enable this test case.

File: build/build.go [hunk 11]
```diff
@@ -394,15 +395,30 @@ func (p *PackageData) InternalBuildContext() *build.Context {
 	return p.bctx
 }
 
+func joinEmbedPatternPos(m1, m2 map[string][]token.Position) map[string][]token.Position {
```
[the remark this answers is not in the document]
Comment (visualfc) [reply build/build.go]: fixed

File: build/embed.go (hunk 7 above)
Comment (visualfc) [reply build/embed.go]: fixed

File: build/embed.go (hunk 8 above)
Comment (visualfc) [reply build/embed.go]: fixed

File: build/build.go [hunk 12]
```diff
@@ -676,6 +710,13 @@ func (s *Session) BuildPackage(pkg *PackageData) (*compiler.Archive, error) {
 	if err != nil {
 		return nil, err
 	}
+	embed, err := embedFiles(pkg, fileSet, files)
+	if err != nil {
+		return nil, err
```
[the remark this answers is not in the document]
Comment (visualfc) [reply build/build.go:701]: `gopherjs build/run` go:embed error keep up with `go build/run` go:embed error.  ( pos: msg)
see <https://github.com/visualfc/goembed/blob/main/load_test.go#L85>

example
```
/Users/vfc/demo/vtest/embed/main.go:9:3: misplaced go:embed directive
```
if add context to very long:
```
failed to generate embedded resources for package "main": failed to associate embed patterns with the AST nodes in package "main": /Users/vfc/demo/vtest/embed/main.go:9:3: misplaced go:embed directive
```

File: compiler/natives/src/embed/internal/embedtest/embed_test.go (hunk 10 above)
Comment (visualfc) [reply compiler/natives/src/embed/internal/embedtest/embed_test.go]: fixed

File: build/build.go [hunk 13]
```diff
@@ -548,12 +558,28 @@ func (s *Session) BuildFiles(filenames []string, pkgObj string, cwd string) erro
...
+				return nil, err
+			}
+			infos = append(infos, info)
+		}
+		return
+	}
+	p, err := ctx.Import(".", root, 0)
+	if err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (nevkontakte) [reply build/build.go]: @visualfc can you please confirm I understood this code correctly?

File: build/build.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (nevkontakte) [reply build/build.go:701]: Fair point, then it's ok to leave as-is.

**Review by nevkontakte — APPROVED** [review]
The pull request looks good to me! 🎉

@visualfc could you please squash all commits into one? We should be good to merge then.

**Comment by visualfc** [comment]
> The pull request looks good to me! 🎉
> 
> @visualfc could you please squash all commits into one? We should be good to merge then.

create new PR https://github.com/gopherjs/gopherjs/pull/1153

**Comment by nevkontakte** [comment]
Force-pushing into this PR would have been fine too, but in any case https://github.com/gopherjs/gopherjs/pull/1153 has been merged, so I'm closing  this one. Thank you again for your contribution!

**Pull request closed without merging**